quic/s2n-quic-core/src/sync/worker.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::sync::primitive::{Arc, AtomicUsize, AtomicWaker, Ordering};
+use cache_padded::CachePadded;
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Creates a worker channel with a Sender and Receiver
+pub fn channel() -> (Sender, Receiver) {
+    let state = Arc::new(State::default());
+    let sender = Sender(state.clone());
+    let receiver = Receiver { state, credits: 0 };
+    (sender, receiver)
+}
+
+/// A handle to the receiver side of the worker channel
+///
+/// This handle is used by the worker to wake up when there is work to do.
+pub struct Receiver {
+    state: Arc<State>,
+    credits: usize,
+}
+
+impl Receiver {
+    /// Acquires work to be processed for the Receiver
+    ///
+    /// `None` is returned when there are no more active Senders.
+    #[inline]
+    pub async fn acquire(&mut self) -> Option<usize> {
+        Acquire(self).await
+    }
+
+    /// Polls work to be processed for the receiver
+    ///
+    /// `None` is returned when there are no more active Senders.
+    #[inline]
+    pub fn poll_acquire(&mut self, cx: &mut Context) -> Poll<Option<usize>> {
+        let state = &*self.state;
+
+        macro_rules! acquire {
+            () => {{
+                // take the credits that we've been given by the senders
+                self.credits += state.remaining.swap(0, Ordering::Acquire);
```

Comment:
Do you want AcqRel since swap does a store and load

Reply:
spoke offline: in the swap operation itself is atomic so the load and store happen atomically. We read the credits and reset them to 0. 

The ordering simply implies ordering with other atomic store operations, which is not critical to correct operation and happen in a relaxed manner.

> Note that using [Acquire](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Acquire) makes the store part of this operation [Relaxed](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Relaxed)